File: bishopTests/src/bishopTests/BitLoopTest.java
package bishopTests;

import bishop.base.BitBoard;
import bishop.base.BitLoop;
import bishop.base.Square;
import org.junit.Assert;
import org.junit.Test;

import java.util.HashSet;
import java.util.Set;
import java.util.SplittableRandom;

public class BitLoopTest {
	@Test
	public void testCorrectResult() {
		final SplittableRandom rng = new SplittableRandom(1234);

		for (int i = 0; i < 100000; i++) {
			final Set<Integer> expectedSquares = new HashSet<>();
			long board = BitBoard.EMPTY;

			for (int square = Square.FIRST; square < Square.LAST; square++) {
				if (rng.nextBoolean()) {
					expectedSquares.add(square);
					board |= BitBoard.of(square);
				}
			}

			final Set<Integer> givenSquares = new HashSet<>();

			for (BitLoop loop = new BitLoop(board); loop.hasNextSquare(); )
				givenSquares.add(loop.getNextSquare());

			Assert.assertEquals(expectedSquares, givenSquares);
		}
	}
}

File: native/tests/bishop/bit_loop_test.cpp

#include "../../bishop/base/bit_board.h"
#include "../../bishop/base/bit_loop.h"
#include "../test_base/test_base.h"

#include <random>
#include <set>

using namespace bishop::base;


TEST (testCorrectResult) {
	::std::minstd_rand rng;
	::std::bernoulli_distribution squareDistribution;

	for (int i = 0; i < 100000; i++) {
		::std::set<Square::Type> expectedSquares;
		BitBoard::Type board = BitBoard::EMPTY;

		for (Square::Type square = Square::FIRST; square < Square::LAST; square++) {
			if (squareDistribution(rng)) {
				expectedSquares.insert(square);
				board |= BitBoard::of(square);
			}
		}

		::std::set<Square::Type> givenSquares;

		for (BitLoop loop(board); loop.hasNextSquare(); )
			givenSquares.insert(loop.getNextSquare());

		assertTrue(expectedSquares == givenSquares);
	}
}


TEST_RUNNER_BEGIN
	RUN_TEST(testCorrectResult);
TEST_RUNNER_END

